plaso/parsers/bsm.py

```diff
@@ -71,1118 +36,447 @@ def __init__(self):
-          'Jumping for the next entry.'.format(
-              token_type, file_object.tell()))
-      try:
-        file_object.seek(
-            record_end_offset - file_object.tell(), os.SEEK_CUR)
-      except (IOError, construct.FieldError) as exception:
-        logger.warning(
-            'Unable to jump to next entry with error: {0!s}'.format(exception))
-        return False
+      token_type_string = self._TOKEN_TYPES.get(token_type, 'UNKNOWN')
+      token_values = self._FormatTokenData(token_type, token_data)
+      event_tokens[token_type_string] = token_values
 
-    event_data = BSMEventData()
-    if parser_mediator.operating_system == definitions.OPERATING_SYSTEM_MACOS:
-      # BSM can be in more than one OS: BSD, Solaris and MacOS.
-      # In MacOS the last two tokens are the return status and the trailer.
-      return_value = extra_tokens.get('BSM_TOKEN_RETURN32')
-      if not return_value:
-        return_value = extra_tokens.get('BSM_TOKEN_RETURN64')
-      if not return_value:
-        return_value = 'UNKNOWN'
+      if token_type in (0x27, 0x72):
+        return_token_values = token_values
 
-      event_data.return_value = return_value
+    if token_data.signature != self._TRAILER_TOKEN_SIGNATURE:
+      raise errors.ParseError('Unsupported signature in trailer token.')
 
-    event_data.event_type = event_type
-    event_data.extra_tokens = extra_tokens
-    event_data.offset = record_start_offset
-    event_data.record_length = record_length
+    if token_data.record_size != header_record_size:
+      raise errors.ParseError(
+          'Mismatch of event record size between header and trailer token.')
 
-    # TODO: check why trailer was passed to event in original while
-    # event was expecting record length.
-    # if extra_tokens:
-    #   trailer = extra_tokens.get('BSM_TOKEN_TRAILER', 'unknown')
+    event_data = BSMEventData()
+    event_data.event_type = event_type
+    event_data.extra_tokens = event_tokens
+    event_data.offset = header_record_offset
+    event_data.record_length = header_record_size
+    event_data.return_value = return_token_values
 
+    date_time = dfdatetime_posix_time.PosixTimeInMicroseconds(
+        timestamp=timestamp)
     event = time_events.DateTimeValuesEvent(
         date_time, definitions.TIME_DESCRIPTION_CREATION)
     parser_mediator.ProduceEventWithEventData(event, event_data)
 
-    return True
-
-  def _RawToUTF8(self, byte_stream):
-    """Copies a UTF-8 byte stream into a Unicode string.
+  def _ParseToken(self, file_object, file_offset):
+    """Parses a token.
 
     Args:
-      byte_stream (bytes): byte stream containing an UTF-8 encoded string.
+      file_object (dfvfs.FileIO): file-like object.
+      file_offset (int): offset of the token relative to the start of
+          the file-like object.
 
     Returns:
-      str: A Unicode string.
+      tuple[int, object]: token type and token data or None if the token
```

Comment:
What is the advantage? The reason it returns None is so that the caller can distinguish between an unsupported token (non-fatal) and a parsing error (fatal).